### PR TITLE
Feature/redux dev

### PR DESCRIPTION
**Pull request by ryansnorek**
#  Redux Conversion

## Loom recording
https://www.loom.com/share/d325ea625182483fbd6f7ea11bfa7541

## Description
Convert component state into redux for scalability.

Trello card: https://trello.com/c/t9qdvrKh

Specific changes:
### /state/actions/index
1. Add --> authenticateUser 
2. Add --> setUserInfo 
4. Add --> fetchStart, fetchEnd, fetchError
5. Delete redundant code --> getUserProfile

### /state/reducers/userReducer
1. Add --> isFetching
2. Add --> userInfo and auth items

### /pages/Home/HomeContainer
1. Replace useEffect calls with redux action --> authenticateUser

### /NavBarLanding/NavBarLanding
1. Replace piece of state --> user.profile --> user.userInfo

## Contributors
Jared Hall
Sean T
@MrJacobSullivan 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have removed unnecessary comments/console logs from my code
- [x] I have made corresponding changes to the documentation if necessary (optional)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any misspellings
- [x] No duplicate code left within changed files
- [x] Size of pull request kept to a minimum
- [x] Pull request description clearly describes changes made & motivations for said changes

### CODE REVIEW

**Review by cainpalmer — APPROVED** [review]
PR is very clean and easy to read, code is mocked up and clean as well. I see there seems to be multiple items defined but not used. Other than that the code looks great.

File: src/state/actions/index.js [hunk 1]
```diff
@@ -27,3 +52,9 @@ export const getEventTemplateStub2 = event => {
   return { type: MENTEE_EVENT_STUB, payload: event };
 };
 // ADMIN TICKETS--------------------------
+
+// ASYNC MANAGEMENT --------------------------
+export const FETCH_START = 'FETCH_START';
```
Comment (MrJacobSullivan) [inline src/state/actions/index.js:57]: This will be useful for handling loading state, nice work!

File: src/state/actions/index.js [hunk 2]
```diff
@@ -3,8 +3,33 @@
...
+      .getUser()
+      .then(async info => {
+        const role_id = await getRole(info.sub);
+        dispatch(setUserInfo({ ...info, role: role_id }));
+      })
+      .then(() => {
+        dispatch(getUserProfile());
+      })
+      .catch(err => {
+        console.log(err);
```
Comment (MrJacobSullivan) [inline src/state/actions/index.js:24]: Could we handle this error in state? That way, if there's an error connecting to the backend, we could let the user know in a friendly way.

File: src/state/reducers/userReducer.js [hunk 3]
```diff
@@ -1,14 +1,26 @@
 // export const USER_EVENT_START = 'USER_EVENT_START';
-import { USER_PROFILE } from '../actions/index';
+import * as a from '../actions/index';
```
Comment (MrJacobSullivan) [inline src/state/reducers/userReducer.js:2]: Just a pet peeve of mine, but could we rename `a` to `ACTIONS` or something? Single-letter variables can get difficult to keep track of. Especially down the road when our redux store is significantly larger than it is now.

File: src/state/reducers/userReducer.js [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
 
 const initialState = {
+  authState: {},
+  authService: {},
+  userInfo: null,
+  isFetching: false,
   photo: '',
   profile: '',
```
Comment (MrJacobSullivan) [inline src/state/reducers/userReducer.js:10]: At this point, can we delete `photo` and `profile` from state? It seems all profile data should be handled in `userInfo` and we aren't currently storing photos of users.

**Review by MrJacobSullivan — CHANGES_REQUESTED** [review]
This is really good work! I especially like the fact that we will now be tracking when state is pending. I think there are a few changes we could make in this PR. It would be very helpful if we tracked errors in application state.

**Review by MrJacobSullivan — APPROVED** [review]
Thank you for adding error handling to the application state! This is a great foundation from which we can build our redux store.

File: src/components/pages/Home/HomeContainer.js [hunk 5]
```diff
@@ -1,42 +1,18 @@
...
-        return setUserInfo(null);
-      });
-    return () => (isSubscribed = false);
-  }, [memoAuthService]);
-
-  useEffect(() => {
-    getUserProfile(userInfo);
-  });
+    dispatch(authenticateUser(authService));
+  }, [dispatch, authService]);
```
Comment (paulstgermain) [inline src/components/pages/Home/HomeContainer.js:15]: _Love_ seeing this shortening and sweetening of our code here - keep it up!

File: src/components/pages/NavBarLanding/NavBarLanding.js [hunk 6]
```diff
@@ -19,6 +19,8 @@ function NavBarLanding(props) {
     'https://joeschmoe.io/api/v1/random'
   );
 
+  console.log(props);
```
Comment (paulstgermain) [inline src/components/pages/NavBarLanding/NavBarLanding.js:22]: Let's make a note to remove this `console.log()` in the future! 😉

File: src/state/actions/index.js [hunk 7]
```diff
@@ -4,13 +4,38 @@
...
+
+// NOTE: appears to be redundant code, commenting out for now
+// user profile is stored in userReducer state labeled "userInfo"
+
+// export const getUserProfile = () => {
+//   return dispatch => {
+//     const token = localStorage.getItem('okta-token-storage');
+//     dispatch(fetchUserProfile(JSON.parse(token).idToken.claims));
+//   };
+// };
```
Comment (paulstgermain) [inline src/state/actions/index.js:38]: Let's make a decision on this commented-out code ASAP!

**Comment by paulstgermain** [comment]
Additional note I'd like to make here - this is a textbook example of what we're looking from in our PR descriptions. Great work and detail here!

**Pull request merged**